consensus/tendermint/messages.go

```diff
@@ -0,0 +1,126 @@
+package tendermint
+
+// Todo: Signature over the messages needs to be handled somewhere. There are 2 options:
+//	1. Add the signature to each message and extend the Validator Set interface to include VerifyMessageSignature
+//	method.
+//	2. The P2P layer signs the message before gossiping to other validators and verifies the signature before passing
+//	the message to the consensus engine.
+//  The benefit of P2P layer handling the verification of the signature is the that the consensus layer can assume
+//  the message is from a validator in the validator set. However, this means that the P2P layer would need to be aware
+//  of the validator set and would need access to the blockchain which may not be a good idea.
+
+type Message[V Hashable[H], H Hash, A Addr] interface {
+	Proposal[V, H, A] | Prevote[H, A] | Precommit[H, A]
+}
+
+type Proposal[V Hashable[H], H Hash, A Addr] struct {
+	Height     uint
+	Round      uint
+	ValidRound *uint
+	Value      *V
+
+	Sender A
+}
+
+type Prevote[H Hash, A Addr] struct {
+	Vote[H, A]
+}
+
+type Precommit[H Hash, A Addr] struct {
+	Vote[H, A]
+}
+
+type Vote[H Hash, A Addr] struct {
+	Height uint
+	Round  uint
+	ID     *H
+
+	Sender A
+}
+
+// messages keep tracks of all the proposals, prevotes, precommits by creating a map structure as follows:
+// height->round->address->[]Message
+
+// Todo: would the following representation of message be better:
+//
+//	  height -> round -> address -> ID -> Message
+//	How would we keep track of nil votes? In golan map key cannot be nil.
+//	It is not easy to calculate a zero value when dealing with generics.
+type messages[V Hashable[H], H Hash, A Addr] struct {
+	proposals  map[uint]map[uint]map[A][]Proposal[V, H, A]
+	prevotes   map[uint]map[uint]map[A][]Prevote[H, A]
+	precommits map[uint]map[uint]map[A][]Precommit[H, A]
+}
+
+func newMessages[V Hashable[H], H Hash, A Addr]() messages[V, H, A] {
+	return messages[V, H, A]{
+		proposals:  make(map[uint]map[uint]map[A][]Proposal[V, H, A]),
+		prevotes:   make(map[uint]map[uint]map[A][]Prevote[H, A]),
+		precommits: make(map[uint]map[uint]map[A][]Precommit[H, A]),
+	}
+}
+
+// Todo: ensure duplicated messages are ignored.
```

Comment:
What’s the difference between a duplicate message and a malicious equivocating message?

Reply:
Equivocated messages are when a validator sends conflicting messages for example a validator can not send both prevote(v) and prevote(nil) for the same height and round. Duplicated messages are when a message is received multiple times this can happen due to how gossiping. Therefore, a prevote(v) for a height and round can be received multiple times.

Reply:
Makes sense, thanks!